gateway/ln-gateway/src/lib.rs

```diff
@@ -109,31 +109,41 @@ pub enum LightningMode {
 
 #[derive(Debug, Error)]
 pub enum GatewayError {
-    #[error("Lightning rpc operation error: {0:?}")]
-    LnRpcError(#[from] tonic::Status),
     #[error("Federation error: {0:?}")]
     FederationError(#[from] FederationError),
     #[error("Other: {0:?}")]
-    Other(#[from] anyhow::Error),
-    #[error("Failed to fetch route hints")]
-    FailedToFetchRouteHints,
+    ClientStateMachineError(#[from] anyhow::Error),
     #[error("Failed to open the database: {0:?}")]
     DatabaseError(anyhow::Error),
     #[error("Federation client error")]
-    ClientNgError,
-}
-
-impl GatewayError {
-    pub fn other(msg: String) -> Self {
-        error!(msg);
-        GatewayError::Other(anyhow!(msg))
-    }
+    LightningRpcError(#[from] LightningRpcError),
+    #[error("Outgoing Payment Error {0:?}")]
+    OutgoingPaymentError(#[from] OutgoingPaymentError),
+    #[error("Invalid Metadata: {0}")]
+    InvalidMetadata(String),
+    #[error("Unexpected state: {0}")]
+    UnexpectedState(String),
 }
 
 impl IntoResponse for GatewayError {
     fn into_response(self) -> Response {
-        let mut err = Cow::<'static, str>::Owned(format!("{self:?}")).into_response();
-        *err.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+        // For privacy reasons, we do not return too many details about the failure of
+        // the request back to the client to prevent malicious clients from
+        // deducing state about the gateway/lightning node.
+        let (error_message, status_code) = match self {
+            GatewayError::OutgoingPaymentError(_) => (
+                format!(
+                    "Error while paying lightning invoice. Outgoing contract will be refunded."
+                ),
+                StatusCode::CONFLICT,
+            ),
+            _ => (
+                format!("An internal gateway error occurred"),
+                StatusCode::INTERNAL_SERVER_ERROR,
+            ),
+        };
+        let mut err = Cow::<'static, str>::Owned(error_message).into_response();
+        *err.status_mut() = status_code;
```

Comment:
@wbobeirne do you have any opinions on what status code should be returned here. Right now, I'm just returning 409 "conflict" when there's an error paying an invoice and 500 for everything else that might go wrong. But I'm happy to change this is there's something the makes more sense on the frontend

Reply:
Nah anything 4XX is fine if we're being intentionally vague. My only thought would be that just a straight 400 is the most generic whereas someone might interpret a 409 to mean that there's some stateful reason that the request doesn't work now, but may have worked in the past or work in the future. 400 tells the client not to try again with that request.

Reply:
Done. Switched to 400